Ex S2:   There are three rods(rod A, rod B, and rod C)and 20 disks of different sizes, which can slide onto any rod. The 20 disk sare placed in ascending order of size on rodA, while the smallest at the top, and the largest at the bottom. Question: Please calculate the minimum number of steps requiredto move the 20 disksfrom rod A to rod C, obeying the following rules:

In [5]:
Hanoi <- function(n,A,B,C) {
    
    if (n == 1) {
        cat("From rod ",A," to rod ",C,"\n")
        
    } else {
        Hanoi(n-1, A, C, B)
        cat("From rod ",A," to rod ",C,"\n")
        Hanoi(n-1, B, A, C)
    }


}



In [11]:
Hanoi(4,"A","B","C")

From rod  A  to rod  B 
From rod  A  to rod  C 
From rod  B  to rod  C 
From rod  A  to rod  B 
From rod  C  to rod  A 
From rod  C  to rod  B 
From rod  A  to rod  B 
From rod  A  to rod  C 
From rod  B  to rod  C 
From rod  B  to rod  A 
From rod  C  to rod  A 
From rod  B  to rod  C 
From rod  A  to rod  B 
From rod  A  to rod  C 
From rod  B  to rod  C 


If we do this recursive function with 20 disk, then it will crush the program. Therefore, I count the numbers of steps in another function which were called count_step. 

In [6]:
count_step <- function (n){
    for(i in c(1:n)) {
        if (i == 1) {
            k = i     
        } else {     
            k = k * 2 + 1       
        }
    }
    cat(k)
}  

In [12]:
count_step(20)

1048575

Minimum number of steps required to move the 20 disks from rod A to rod C is 1048575.

Ex S3:   Please use the “apply” command to calculate the medians, maximums, and minimums, of each row and each column. Missing values could be ignored.

In [14]:
x <-matrix(c(3600, 5000, 12000, NA, 1000, 2000, 600, 7500, 1800, 9000,3600, 4500, 
             10000, 8500, 3000, 10000, 1000, NA, 1200, 10000,3800, 5500, 9000, 6000, 
             6600, 3000, 9600, 6500, 8200, 8000,5000, 6600, 13000, 4500, 5000, NA, 10600, 
             9500, 7600, 6000,6600, 8000, 17000, 3000, 7000, 1000, 12600, 8500, 6000, NA),
           5,10, byrow = TRUE)
x

[,1] [,2] [,3]  [,4] [,5] [,6]  [,7]  [,8] [,9] [,10]
[1,] 3600 5000 12000   NA 1000  2000   600 7500 1800  9000
[2,] 3600 4500 10000 8500 3000 10000  1000   NA 1200 10000
[3,] 3800 5500  9000 6000 6600  3000  9600 6500 8200  8000
[4,] 5000 6600 13000 4500 5000    NA 10600 9500 7600  6000
[5,] 6600 8000 17000 3000 7000  1000 12600 8500 6000    NA

By row:

median

In [15]:
apply(x,1,median,na.rm=TRUE)

[1] 3600 4500 6550 6600 7000

maximums

In [16]:
apply(x,1,max,na.rm=TRUE)

[1] 12000 10000  9600 13000 17000

minimums

In [17]:
apply(x,1,min,na.rm=TRUE)

[1]  600 1000 3000 4500 1000

By column:

median

In [18]:
apply(x,2,median,na.rm=TRUE)

[1]  3800  5500 12000  5250  5000  2500  9600  8000  6000  8500

maximums

In [19]:
apply(x,2,max,na.rm=TRUE)

[1]  6600  8000 17000  8500  7000 10000 12600  9500  8200 10000

minimums

In [27]:
apply(x,2,min,na.rm=TRUE)

[1] 3600 4500 9000 3000 1000 1000  600 6500 1200 6000

Ex S8: Please find x, y, and z.

In [17]:
A <- matrix(c(1,-3,1,1,-2,3,1,-1,1),3,3,byrow=TRUE)
b <- c(4,6,4)

In [18]:
solve(A,b)

[1] 3 0 1

x=3; y=0; z=1